lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,310 @@
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidClientError = require('../errors/invalid-client-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var InvalidScopeError = require('../errors/invalid-scope-error');
+var OAuthError = require('../errors/oauth-error');
+var Promise = require('bluebird');
+var Request = require('../request');
+var Response = require('../response');
+var ServerError = require('../errors/server-error');
+var UnauthorizedClientError = require('../errors/unauthorized-client-error');
+var is = require('../validator/is');
+var tokenUtil = require('../utils/token-util');
+var url = require('url');
+
+/**
+ * Response types.
+ */
+
+var responseTypes = {
+  code: require('../response-types/code-response-type'),
+  token: require('../response-types/token-response-type')
+};
+
+/**
+ * Constructor.
+ */
+
+function AuthorizeHandler(options) {
+  options = options || {};
+
+  if (!options.authorizationCodeLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `authorizationCodeLifetime`');
+  }
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getClient) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getClient()`');
+  }
+
+  if (!options.model.saveAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveAuthorizationCode()`');
+  }
+
+  this.authorizationCodeLifetime = options.authorizationCodeLifetime;
+  this.authenticateHandler = new AuthenticateHandler(options);
+  this.model = options.model;
+}
+
+/**
+ * Authorize Handler.
+ */
+
+AuthorizeHandler.prototype.handle = function(request, response) {
+  if (!(request instanceof Request)) {
+    throw new InvalidArgumentError('Invalid argument: `request` must be an instance of Request');
+  }
+
+  if (!(response instanceof Response)) {
+    throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Response');
+  }
+
+  if (false === request.body.allowed) {
+    return Promise.reject(new AccessDeniedError('Access denied: user denied access to application'));
+  }
+
+  var fns = [
+    this.generateAuthorizationCode(),
+    this.getAuthorizationCodeLifetime(),
+    this.getClient(request),
+    this.getUser(request, response)
+  ];
+
+  return Promise.all(fns)
+    .bind(this)
+    .spread(function(authorizationCode, expiresAt, client, user) {
+      var scope;
+      var state;
+
+      return Promise.bind(this)
+        .then(function() {
+          scope = this.getScope(request);
+          state = this.getState(request);
+
+          return this.saveAuthorizationCode(authorizationCode, expiresAt, scope, client, user);
+        })
+        .then(function(code) {
+          var responseType = this.getResponseType(request, code);
+          var redirectUri = this.buildSuccessRedirectUri(client.redirectUri, responseType);
```

Comment:
The redirect URI should be validated along with the client before any other processing.